remove 0.8 deprecations from pvsystem.py, modelchain.py, rework temperature model param deprecation

wholmgren opened the pull request:
- [x] Closes #1030 closes #966 
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

I'm still struggling with the right approach for #1030. Here are examples of the behavior in this PR:

```python
In [109]: PVSystem()
/Users/holmgren/git_repos/pvlib-python/pvlib/pvsystem.py:219: pvlibDeprecationWarning: temperature_model_parameters, racking_model, and module_type are not specified. Reverting to deprecated default: SAPM cell temperature model parameters for a glass/glass module in open racking. In v0.9 PVSystem.temperature_model_parameters will be empty dict unless set by temperature_model_parameters or valid combination of racking_model and module_type. Pass non-None value to any of these parameters to silence this warning.
  pvlibDeprecationWarning)
Out[109]:
PVSystem:
  name: None
  surface_tilt: 0
  surface_azimuth: 180
  module: None
  inverter: None
  albedo: 0.25
  racking_model: None
  module_type: None
  temperature_model_parameters: {'a': -3.47, 'b': -0.0594, 'deltaT': 3}

In [110]: PVSystem(racking_model=False)
Out[110]:
PVSystem:
  name: None
  surface_tilt: 0
  surface_azimuth: 180
  module: None
  inverter: None
  albedo: 0.25
  racking_model: False
  module_type: None
  temperature_model_parameters: {}
```

This change also causes 28 new warnings from just `test_pvsystem.py`. I'm sure there are more in `test_modelchain.py`. In practice, I think this only matters to users that call `PVSystem.sapm_celltemp`, so we could push the deprecation down to this method.

wholmgren commented:
Moving the deprecation warning from `PVSystem.__init__` to `PVSystem.sapm_celltemp` works with one caveat: `ModelChain.infer_temperature_method` cannot infer the SAPM method. A temporary work around could be:

```python
    def infer_temperature_model(self):
        params = set(self.system.temperature_model_parameters.keys())
        # revert this change in 0.9
        if set(['a', 'b', 'deltaT']) <= params or (if not params and self.system.racking_model is None and self.system.module_type is None):
            return self.sapm_temp  # system.sapm_celltemp will assign default keys and emit warning at runtime
```

Does that all seem reasonable @cwhanse?

cwhanse commented on pvlib/pvsystem.py:
```diff
@@ -201,20 +201,21 @@ def __init__(self,
+                'temperature_model_parameters, racking_model, and module_type '
+                'are not specified. Reverting to deprecated default: SAPM '
+                'cell temperature model parameters for a glass/glass module '
+                'in open racking. In v0.9 '
+                'PVSystem.temperature_model_parameters will be empty dict '
+                'unless set by temperature_model_parameters or valid '
+                'combination of racking_model and module_type. Pass non-None '
```
```suggestion
                'in open racking. In v0.9, temperature_model_parameters or a valid '
                'combination of racking_model and module_type will be required. '
                ' Pass non-None '
```

I'd prefer to be more specific how to use v0.9.

wholmgren replied:
I'm ok with "will be required" if the warning is emitted from `PVSystem.sapm_celltemp`. I don't like it if the warning is emitted from the `PVSystem.__init__` because it suggests a positional argument.

cwhanse replied:
> `PVSystem()` has always been a valid, non-warned statement. There's never been a requirement that `PVSystem` is a complete description of a system according to some model. We can discuss changing that, but it's inconsistent to require temperature parameters and nothing else.

Agree, and point taken. I'm OK with the change proposed for `ModelChain.infer_temperature_model`.

wholmgren replied:
changed the text in the warning in the method

cwhanse commented:
> Moving the deprecation warning from `PVSystem.__init__` to `PVSystem.sapm_celltemp` works with one caveat: `ModelChain.infer_temperature_method` cannot infer the SAPM method. A temporary work around could be:

I'm not following here - why move the warning from `PVSystem.__init__` to `sapm_celltemp`? I think it works better to detect the None condition and apply the deprecated default when PVSystem is initialized, and let ModelChain assume that PVSystem has taken care of that.

wholmgren commented:
`PVSystem()` has always been a valid, non-warned statement. There's never been a requirement that `PVSystem` is a complete description of a system according to some model. We can discuss changing that, but it's inconsistent to require temperature parameters and nothing else.

wholmgren commented:
I think this is now ready to merge.

Sorry for the churn as I worked out inconsistencies in flake8 settings between stickler and running locally. Added a flake8 section to `setup.cfg` so that running `flake8` locally will use the same configuration as stickler. Added ignore statements for W503 and W504 (break before/after operator) to be consistent with current best practice.

cwhanse commented on pvlib/pvsystem.py:
```diff
@@ -170,12 +170,12 @@ class PVSystem(object):
                  surface_tilt=0, surface_azimuth=180,
                  albedo=None, surface_type=None,
-                 module=None, module_type='glass_polymer',
+                 module=None, module_type=None,
```
I think changing the default to None should be mentioned in whatsnew

wholmgren replied:
good catch. added a few awkward sentences to `API Changes with Deprecations`

cwhanse commented on docs/sphinx/source/whatsnew/v0.8.0.rst:
```diff
@@ -12,6 +12,14 @@ API Changes with Deprecations
+* :py:class:`pvlib.pvsystem.PVSystem` ``module_type`` and ``racking_model`` now
+  default to ``None``. This continues a deprecation of assuming SAPM values
+  for cell temperature modeling. In this v0.8 release series, calling
+  :py:meth:`pvlib.pvsystem.PVSystem.sapm_celltemp` will cause
+  ``PVSystem.temperature_model_parameters`` to be set to SAPM values for a
+  glass/glass module in open racking and emit a warning. In v0.9, users must
```
```suggestion
  :py:meth:`pvlib.pvsystem.PVSystem.sapm_celltemp` without ``PVSystem.temperature model parameters``,
  or a valid combination of ``PVsystem.module_type`` and ``PVsystem.racking_model``, will cause
  ``PVSystem.temperature_model_parameters`` to be set to SAPM values for a
  glass/glass module in open racking and emit a warning. In v0.9, users must
```

Not less awkward but maybe more clear

kanderso-nrel commented:
FYI I noticed there's still an entry to `sapm_aoi_loss` in api.rst: https://pvlib-python.readthedocs.io/en/latest/api.html#sandia-array-performance-model-sapm
